apps/slackbot/task_agent.py

```diff
@@ -87,6 +89,169 @@ def from_llm_and_tools(
+                result = result["command"]["args"]["response"]
+            else:
+                # print(result)
+                result = str(result)
+        except json.JSONDecodeError:
+            result = assistant_reply
+
+        return self.process_output(result)
+        
+    def update_logger(self, assistant_reply):
+        logger_step = {"Step": f"{self.loop_count}/{self.max_iterations}"} 
+        try:
+            reply_json = json.loads(assistant_reply)
+            logger_step["reply"] = reply_json
+        except json.JSONDecodeError:
+            logger_step["reply"] = assistant_reply  # last reply is a string
+        self.command_log.append(logger_step)
+        return None 
+        
+    def reformulate_action(self, task, assistant_reply):
+        
+        action = self.output_parser.parse(assistant_reply)
+        print("action:", action)
+        if action == self.previous_action:
+            if action.name == "Search" or action.name == "Context Search":
+                print(
+                    "Action name: ", action.name, "\nStart reformulating the query"
+                )
+                instruction = (
+                    f"You want to search for useful information to answer the query: {task}."
+                    f"The original query is: {action.args['query']}"
+                    f"Reformulate the query so that it can be used to search for relevant information."
+                    f"Only return one query instead of multiple queries."
+                    f"Reformulated query:\n\n"
+                )
+                openai.api_key = cfg.OPENAI_API_KEY
+                response = openai.Completion.create(
+                    engine="text-davinci-003",
+                    prompt=" ".join(str(i) for i in self.previous_message)
+                    + "\n"
+                    + instruction,
+                    temperature=0,  # make this low for predictable outcome
+                    max_tokens=1024,
+                    top_p=1,
+                    frequency_penalty=0,
+                    presence_penalty=0,
+                )
+                reformulated_query = response["choices"][0]["text"]
+                action.args["query"] = reformulated_query
+                
+        return action
+
+    def observations_from_actions(self, action):
+        """_summary_
+
+        Args:
+            action (_type_): _description_
+
+        Returns:
+            Result: Formatted return message
+        """
+        tools = {t.name: t for t in self.tools}
+        if action.name == "finish":
+            self.loop_count = self.max_iterations
+            result = "Finished task. "
+        elif action.name in tools:
+            tool = tools[action.name]
+
+            if tool.name == "UserInput":
+                return {"type": "user_input", "query": action.args["query"]}
+
+            try:
+                observation = tool.run(action.args)
+            except ValidationError as e:
```

Comment:
Same comment as earlier about logging exceptions instead of silently swallowing them,.